### DISCUSSION TITLE
Macro issue

### DISCUSSION THREAD

**Discussion by gimmemabrewski** [X3n0m0rph59/eruption#107] [Q&A] [answered]
I'm setting up some macros and running into an issue. Perhaps what I'm trying to do isn't possible but here's the jist of it... I wrote a zsh script that uses xclip to take copied text and "wrap it" with different characters. This would be useful to me for writing scripts. For example, I can copy a group of text, and the script will quote each line of text. Anyway, the issue I'm running into is running the script directly from lua. I'm not familiar with lua at all but it seems similar to python, which I only just started learning lol. Here's the segment from my lua script.
```lua

		-- place your code here --
		debug("Executing: 'single quote macro'")

		-- cut selected text
		inject_key_with_delay(29, true, 0)        -- left ctrl down
		inject_key_with_delay(45, true, 5)       -- 'x' down
		inject_key_with_delay(45, false, 10)     -- 'x' up
		inject_key_with_delay(29, false, 15)     -- left ctrl up

		-- run 'text-macro --wrap -q'
-- 		inject_key_with_delay(56, true, 45)       -- left alt down
-- 		inject_key_with_delay(29, true, 50)        -- left ctrl down
-- 		inject_key_with_delay(125, true, 55)       -- left meta down

-- 		inject_key_with_delay(16, true, 60)       -- 'q' down
-- 		inject_key_with_delay(16, false, 65)       -- 'q' up
		os.execute("/home/myusername/.local/bin/text-macros --wrap -q")
-- 		inject_key_with_delay(56, false, 70)       -- left alt up
-- 		inject_key_with_delay(29, false, 70)        -- left ctrl up
-- 		inject_key_with_delay(125, false, 70)       -- left meta up

		-- paste selected text
		inject_key_with_delay(29, true, 200)        -- left ctrl down
		inject_key_with_delay(47, true, 225)       -- 'v' down
		inject_key_with_delay(47, false, 230)     -- 'v' up
		inject_key_with_delay(29, false, 235)     -- left ctrl up

```
I get an error from eruption in my logs `Error: Can't open display: (null)`. If I comment the `os.execute()` line and uncomment the others it works but it requires me setting up a bunch of keyboard shortcuts. I'd much prefer to run the shell script directly if that's possible.  I've tried moving the script to /usr/bin and running it with `io.popen()` gives the same results. If I run a lua shell and run the script with either `os.execute()` or `io.popen()` it works just fine so I'm guessing it's a library (requires?) issue, permissions issue, or systemd problem :man_shrugging:  Any help would be greatly appreciated! Btw, I'm loving the eruption driver. Nice work!

PS  I tried messing with the timings as well in case I wasn't giving enough time for the command to execute. No luck obviously

**Comment by X3n0m0rph59** [comment, marked as answer]
Hi @gimmemabrewski! Thanks for posting this question!

I believe you need to provide the correct environment variables to your shell script. Additionally it would be better to not run it as the user the `eruption` daemon runs as (`root`), but as your user instead.

Please replace `USER`, `USER_ID` and `DISPLAY` with the correct values for your system:

File `macro.sh`:
```bash
#!/bin/bash
USER=`whoami`
USER_ID=`id -u`
DISPLAY=":1"

DBUS_SESSION_BUS_ADDRESS=unix:path=/run/$USER/$USER_ID/bus

export DISPLAY=$DISPLAY
export DBUS_SESSION_BUS_ADDRESS=$DBUS_SESSION_BUS_ADDRESS

/usr/bin/notify-send "Message from Eruption"
```
> The file `macro.sh` needs to be executable. (`chmod +x macro.sh`)

You may want to replace the line `/usr/bin/notify-send "Message from Eruption"` with your own command invocations like `/usr/bin/xclip`.

Please replace `user` with the correct value for your system:
Somewhere in the Lua script file:
```lua
result = system("/usr/bin/sudo", {"-u", "user", "/home/user/.local/bin/macro.sh"})
if result ~= 0 then
    error("Command execution failed with result: " .. result)
end
```

**Reply by gimmemabrewski** [reply]
It works! Thank you!! I went about it slightly different (used your example as a 'launcher script'). I'm gonna make a few more adjustments and I'll share everything here for anyone that might be interested :stuck_out_tongue:

**Reply by gimmemabrewski** [reply]
I did have another question. If I don't want my macros to be profile specific, does that mean I need to add the configuration stanza for each macro to every .profile file? I'm referring to this...
```
[[config.Macros]]
type = 'string'
name = 'requires'
value = 'macros/mygame'
``` 
??  Thanks again

**Reply by X3n0m0rph59** [reply]
@gimmemabrewski

> I did have another question. If I don't want my macros to be profile specific, does that mean I need to add the configuration stanza for each macro to every .profile file?

Yes, that's correct. But there can only be one `requires` stanza per `.profile` file. A solution would be to simply copy your favorite `.profile` file, change the UUID and adapt it to suit your gaming needs by including `macros/mygame` instead of the default macros from the file `macros/user-macros.lua`. Eruption ships with an example for this (`Gaming: Starcraft 2` profile which requires `macros/starcraft2` instead of the default `macros/user-macros`).

By using one or multiple `.profile` files per game, you could probably also use the automatic profile switching mechanism that is provided by `eruption-process-monitor` daemon.

I think it would be nice to have a working mechanism to switch the macro presets of the active profile on the fly like this:
```sh
$ eruptionctl param "Macros" requires 'macros/special-macros'
```

and later on, switch back to the default presets:
```sh
$ eruptionctl param "Macros" requires 'macros/user-macros'
```

This currently crashes the daemon. I will look into this...

**Reply by gimmemabrewski** [reply]
> But there can only be one requires stanza per .profile file. 

Yes, I figured this out :laughing: . I thought at first I needed a lua script for every macro before realizing I could just add to the if else statement :man_facepalming: . 

> A solution would be to simply copy your favorite .profile file, change the UUID and adapt it to suit your gaming needs by including macros/mygame instead of the default macros from the file macros/user-macros.lua

I don't game but I might use this tip for other needs :+1: 

Btw, my script got to be a lot more than just macros lol I'll probably take some sections and share it as an example in another thread. Thanks again!
